refactor: use a refinement type for Data.Vec (#586)
plt-amy/cubical-1lab: src/Data/Vec/Base.lagda.md

open import 1Lab.Type

open import Data.Product.NAry
open import Data.Maybe.Base

> open import Data.List.Base hiding (head ; tail ; lookup) renaming (tabulate to tabulateL ; _++_ to _++L_ ; zip to zipL ; zip-with to zip-withL)

Reviewer: it might be a better idea to import this qualified

-->

```agda
data Length {ℓ} {A : Type ℓ} : List A → Nat → Type ℓ where
  zero : Length [] zero
  suc  : ∀ {x xs n} → Length xs n → Length (x ∷ xs) (suc n)

instance
  Length-zero : Length {A = A} [] zero
  Length-zero = zero

  Length-suc : ∀ {x n} {xs : List A} → ⦃ _ : Length xs n ⦄
    → Length (x ∷ xs) (suc n)
  Length-suc ⦃ l ⦄ = suc l

  Length-length : ∀ {xs : List A} → Length xs (length xs)
  Length-length {xs = []} = zero
  Length-length {xs = x ∷ xs} = suc Length-length
  {-# INCOHERENT Length-length #-}

  Length-dec : ∀ {xs : List A} → Dec (Length xs n)
  Length-dec {n = zero} {xs = []} =  yes $ Length-zero
  Length-dec {n = suc n} {xs = x ∷ xs} =  caseᵈ Length xs n of λ where
    (yes l) -> yes $ Length-suc ⦃ l ⦄

>     (no ¬l) -> no λ { (suc l) → ¬l l }

Reviewer: ```suggestion
    (yes l) → yes $ Length-suc ⦃ l ⦄
    (no ¬l) → no λ { (suc l) → ¬l l }
```

  Length-dec {n = suc n} {xs = []} = no λ ()
  Length-dec {n = zero} {xs = x ∷ xs} = no λ ()

length-uncons : ∀ {xs n x} → Length {A = A} (x ∷ xs) (suc n) → Length xs n
length-uncons (suc l) = l

> has-lengthᵢ : ∀ {xs : List A} {n : Nat} → (Length xs n) → length xs ≡ᵢ n

Reviewer: ```suggestion
has-lengthᵢ : ∀ {xs : List A} {n : Nat} → Length xs n → length xs ≡ᵢ n
```
missed one

has-lengthᵢ {xs = []} {n = zero} len = reflᵢ
has-lengthᵢ {xs = x ∷ xs} {n = suc n} len = apᵢ suc $ has-lengthᵢ $ length-uncons len

has-length : ∀ {xs : List A} {n : Nat} → Irr (Length xs n) → length xs ≡ n

> has-length l = Id≃path.to $ has-lengthᵢ (recover l)

Reviewer: all of this should be lifted to a different module like `Data.List.Length`

record Vec {ℓ} (A : Type ℓ) (n : Nat) : Type ℓ where
  constructor vec
  field
    lower   : List A
    ⦃ len ⦄ : Irr (Length lower n)

pattern []v = vec []

infixr 20 _∷v_
_∷v_ : ∀ {n} → A → Vec A n → Vec A (suc n)
_∷v_ v (vec vs ⦃ p ⦄) = vec (v ∷ vs) ⦃ suc <$> p ⦄

data Vec-view {ℓ} {A : Type ℓ} : {n : Nat} → Vec A n → Type ℓ where
  []     : Vec-view {n = 0} []v

>   _∷_  : ∀ {n} (a : A) → (vs : Vec A n) → Vec-view {n = suc n} (a ∷v vs)

Reviewer: ```suggestion
data Vec-view {ℓ} {A : Type ℓ} : {n : Nat} → Vec A n → Typeω where
  []   : Vec-view []v
  _∷_  : ∀ {n} a (vs : Vec A n) → Vec-view (a ∷v vs)
```

putting it in `Typeω` is a performance hack to reduce the time we spend in the coverage checker. because `Vec-view` is an indexed type it has formal Kan operations to handle whenever it is pattern-matched on but, if you put it in `Typeω` instead, those aren't generated. for now.

vec-view : ∀ {n} (v : Vec A n) → Vec-view v
vec-view {n = 0} []v = []
vec-view {n = suc k} (vec (x ∷ xs) ⦃ l ⦄) =
  x ∷ vec xs ⦃ forget $ length-uncons $ recover l  ⦄
vec-view {n = zero} (vec (x ∷ xs) ⦃ l ⦄) with () <- recover l

> vec-view {n = suc n} (vec [] ⦃ l ⦄) with () <- recover l

Reviewer: ```suggestion
vec-view : ∀ {n} (v : Vec A n) → Vec-view v
vec-view {n = zero}  (vec [] ⦃ l ⦄) = []
vec-view {n = zero}  (vec (x ∷ xs) ⦃ l ⦄) with () ← recover l
vec-view {n = suc k} (vec (x ∷ xs) ⦃ l ⦄) = x ∷ vec xs ⦃ length-uncons <$> l ⦄
vec-view {n = suc n} (vec [] ⦃ l ⦄) with () ← recover l
```

list→vec : (xs : List A) → Vec A (length xs)
list→vec xs = vec xs

head : Vec A (suc n) → A
head (vec (x ∷ xs)) = x
head (vec [] ⦃ l ⦄) with () ← recover l

tail : Vec A (suc n) → Vec A n
tail v with (x ∷ xs) ← vec-view v = xs
```

The type `Vec A n` [is equivalent to] the type $\rm{Fin}(n) \to A$, i.e.,
the functions from the [[standard finite set]] with $n$ elements to the
type $A$. The halves of this equivalence are called `lookup`{.Agda} and
`tabulate`{.Agda}.

[is equivalent to]: Data.Vec.Properties.html

```agda
lookup : Vec A n → Fin n → A
lookup (vec xs ⦃ len ⦄) (fin n ⦃ n<xs ⦄) =

>   from-just! _ $ !?-just xs n $ ≤-trans n<xs $ subst (Nat._≤ length xs) (has-length $ len) auto

Reviewer: ```suggestion
lookup : Vec A n → Fin n → A
lookup (vec xs) (fin n) = from-just! _ $ !?-just xs n p where abstract
  p : n Nat.< length xs
  p = ≤-trans auto $ subst (Nat._≤ length xs) (has-length auto) auto
```

```

## List syntax {defines="list-syntax-for-vectors"}

A similar type to `Vec`{.Agda} can be defined by _recursion_ as an
iteraded `non-dependent product`{.Agda}. The resulting type `Vecₓ`{.Agda}
has the advantage of supporting usual tuple syntax, but is fiddlier to
eliminate. This is solved by implementing the `From-product`{.Agda}
typeclass for `Vec`{.Agda}, which enables list syntax for the latter.

```agda
instance
  From-prod-Vec : From-product A (Vec A)
  From-prod-Vec .From-product.from-prod = go where
    go : ∀ n → Vecₓ A n → Vec A n
    go zero xs                = []v
    go (suc zero) xs          = xs ∷v []v
    go (suc (suc n)) (x , xs) = x ∷v go (suc n) xs

len-++ : ∀ {ℓ n m} {A : Type ℓ} {xs ys : List A} → Length xs n → Length ys m → Length (xs ++ℓ ys) (n + m)
len-++ zero l' = l'

> len-++ (suc l) l' = Length-suc ⦃ len-++ l l' ⦄

Reviewer: ditto different module

_++_ : ∀ {n k} → Vec A n → Vec A k → Vec A (n + k)

> (vec xs) ++ (vec ys) = vec (xs ++ℓ ys) ⦃ liftA2 len-++ auto auto ⦄

Reviewer: ```suggestion
vec xs ++ vec ys = vec (xs ++ℓ ys) ⦃ liftA2 len-++ auto auto ⦄
```

Vec-elim
  : ∀ {ℓ ℓ'} {A : Type ℓ} (P : ∀ {n} → Vec A n → Type ℓ')
  → P []v
  → (∀ {n} x (xs : Vec A n) → P xs → P (x ∷v xs))
  → ∀ {n} (xs : Vec A n) → P xs
Vec-elim P p[] p∷ v with vec-view v
... | [] = p[]
... | (x ∷ xs) = p∷ x xs $ Vec-elim P p[] p∷ xs
```

<!--
```agda
Vec-cast : {x y : Nat} → x ≡ y → Vec A x → Vec A y
Vec-cast {A = A} {x = x} {y = y} p (vec l ⦃ len ⦄) =
  vec l ⦃ subst (λ n → Length l n) p <$> len ⦄

len-tab : ∀ {n} → (v : Fin n → A) → Length (tabulateℓ v) n
len-tab {n = zero} v = zero

> len-tab {n = suc n} v = suc (len-tab {n = n} (v ∘ fsuc))

Reviewer: ditto separate module

```
-->

```agda
tabulate : (Fin n → A) → Vec A n
tabulate v  = vec (tabulateℓ v) ⦃ forget (len-tab v) ⦄

len-map : ∀ {ℓ ℓ' n} {A : Type ℓ} {B : Type ℓ'} (f : A → B) (xs : List A) → Length xs n → Length (f <$> xs) n
len-map {n = zero} f [] x = zero

> len-map {n = suc n} f (x ∷ xs) (suc l) = suc (len-map f xs l)

Reviewer: ditto separate module

instance
  Map-Vec : ∀ {n} → Map (eff (λ A → Vec A n ) )

>   Map-Vec = record { map = λ { f (vec l ⦃ len ⦄) → vec (f <$> l) ⦃ len-map f l <$> len ⦄ } }

Reviewer: ```suggestion
  Map-Vec .Map.map f (vec l) = vec (f <$> l) ⦃ len-map f l <$> auto ⦄
```

-->

```agda

> 

Reviewer: ```suggestion
```

  From-prod-Vec : From-product A (Vec A)
  From-prod-Vec .From-product.from-prod = go where
    go : ∀ n → Vecₓ A n → Vec A n
    go zero xs                = []v
    go (suc zero) xs          = xs ∷v []v
    go (suc (suc n)) (x , xs) = x ∷v go (suc n) xs

_++_ : ∀ {n k} → Vec A n → Vec A k → Vec A (n + k)
vec xs ++ vec ys = vec (xs ++L ys) ⦃ liftA2 len-++ auto auto ⦄

Vec-elim
  : ∀ {ℓ ℓ'} {A : Type ℓ} (P : ∀ {n} → Vec A n → Type ℓ')
  → P []v
  → (∀ {n} x (xs : Vec A n) → P xs → P (x ∷v xs))
  → ∀ {n} (xs : Vec A n) → P xs
Vec-elim P p[] p∷ v with vec-view v
... | [] = p[]
... | (x ∷ xs) = p∷ x xs $ Vec-elim P p[] p∷ xs
```

<!--
```agda
Vec-cast : {x y : Nat} → x ≡ y → Vec A x → Vec A y
Vec-cast {A = A} {x = x} {y = y} p (vec l ⦃ len ⦄) =
  vec l ⦃ subst (λ n → Length l n) p <$> len ⦄
```
-->

```agda
tabulate : (Fin n → A) → Vec A n
tabulate v  = vec (tabulateL v) ⦃ forget (len-tab v) ⦄

instance
  Map-Vec : ∀ {n} → Map (eff (λ A → Vec A n ) )
  Map-Vec .Map.map f (vec l) = vec (f <$> l) ⦃ len-map f l <$> auto ⦄

zip :  Vec A n → Vec B n → Vec (A × B) n
zip (vec u) (vec v) = vec (zipL u v) ⦃ liftA2 (len-zip _ _) auto auto ⦄

zip-with : (A → B → C) → Vec A n → Vec B n → Vec C n

> zip-with f (vec u) (vec v) = vec (zip-withL f u v) ⦃  liftA2 (len-zip-with _ _ _) auto auto   ⦄

Reviewer: ```suggestion
zip-with f (vec u) (vec v) = vec (zip-withL f u v) ⦃ liftA2 (len-zip-with _ _ _) auto auto ⦄
```

also, line length limit?
